### PR TITLE
[Onboarding - Stage 2] Onboarding Flow - enabled

### PR DESCRIPTION
**Pull request by filip-solecki**
### Fixed Issues

$ https://github.com/Expensify/App/issues/39969
$ https://github.com/Expensify/App/issues/39918
$ https://github.com/Expensify/App/issues/39925
PROPOSAL:

### Tests

1. Log in as a new user 
2. Verify if new Onboarding modal shows up
3. Input a name and a surname in the input fields, press continue button.
4. Click on any of the displayed options and press continue.
5. Verify that a video modal pops up with the app explanation video, if you're on desktop you should also get auto-navigated to the concierge report.
6. Verify that you've received the correct message in the concierge report (it depends on the purpose you've chosen).

**Note:** To test with an existing user, you can initiate the flow from **Settings > Troubleshoot > Onboarding Flow**.

#### With the creation of a workspace (_Track business expenses..., Manage my team’s..._)

1. Open the app with an entire new user.
2. Verify that the "What do you want to do today?" modal appears.
3. Select any tutorial.
4. Click on "Continue".
5. Verify that the "What’s your name?" modal appears.
6. Enter any names.
7. Click on "Continue".
8. Verify that the "Where do you work?" modal appears.
9. Enter any name.
10. Click on "Continue".
11. Verify that the app redirects to the home page. 
12. Open the Concierge chat.
13. Verify that the app has generated the proper messages.
14. Verify that the app has created a workspace with the previously entered name. 
15. Verify that attachment messages are replaced with videos, once the backend sends an update.

#### Without the creation of a workspace (_others_)

1. Open the app with an entire new user.
2. Verify that the "What do you want to do today?" modal appears.
3. Select any tutorial.
4. Click on "Continue".
5. Verify that the "What’s your name?" modal appears.
6. Enter any names.
7. Click on "Continue".
8. Verify that the app redirects to the home page. 
9. Open the Concierge chat.
10. Verify that the app has generated the proper messages.
11. Verify that attachment messages are replaced with videos, once the backend sends an update.

- [x] Verify that no errors appear in the JS console

### Offline tests

1. Log in as a new user 
2. Verify if new Onboarding modal shows up
3. Force offline from Test tool menu
4. Input a name and a surname in the input fields, press continue button.
5. Click on any of the displayed options and press continue.
6. Verify that animation shows instead of video

### QA Steps

Same as tests

- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/Expensify/App/assets/153545991/ccdd035d-ccad-4c41-b01b-008d764df54b

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/Expensify/App/assets/153545991/bcd2038e-fe69-4a7b-8ddc-a4b21ea68946

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/Expensify/App/assets/153545991/02049eab-64f8-4fb9-8665-caf4cd69bd72

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/Expensify/App/assets/153545991/8960d437-755d-4763-ae04-d2b03e1bc2e4

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/App/assets/153545991/a14d9db6-c174-4e1a-ae29-d9cc44fd6883

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/Expensify/App/assets/153545991/719fba01-7860-4d8c-bfc4-07d944547bc4

</details>

### CODE REVIEW

**Review by mountiny — COMMENTED** [review]
The changes look good, lets get this tested well now 🚀

**Comment by mountiny** [comment]
@hoangzinh are you able to do a round of testing on this one today please?

**Comment by filip-solecki** [comment]
Yes, I am investigating this issue, I have to check if new arch caused this issue or not

**Comment by filip-solecki** [comment]
Hmm, it is reproducible also on main in my case, how about you @hoangzinh ?

**Comment by cdOut** [comment]
> @filip-solecki I got an intermittent issue after sign up with a new account, sometimes it displays the onboarding modal, sometimes it shows infinite loading. Do you face same issue as me?

@hoangzinh this issue has started to appear after the new architecture got merged into main, it doesn't only happen on this branch. Here's the steps to fix it:

1. After downloading the dependencies with `npm install` you must run the clean command `npm run clean`
2. When building for web / native web / desktop we need to clear site data by going into `Inspect Element`, into the `Application` tab and then into `Storage`. From this window we need to press the `Clear site data` button and then reload the site / desktop app.
3. For native IOS / Android we also need to clear the app cache by going into `Settings` and `Apps`, selecting the Expensify app and clear it's cache data.

Here's where you can clear the data for web based apps:
Screenshot 2024-04-11 at 13 43 49

**Comment by filip-solecki** [comment]
Hmm, I am not sure if it is code related problem  then. In my case it still exists but exists also on main

**Comment by filip-solecki** [comment]
@hoangzinh check it now, it is not reproducible in my case right now

**Comment by filip-solecki** [comment]
Ok, so I am marking as `Ready for review`

**Comment by trjExpensify** [comment]
Whoops Melv, I think @hoangzinh is supposed to be on the checklist! :)

**Comment by mountiny** [comment]
Making a new build https://github.com/Expensify/App/actions/runs/8691065955

@hoangzinh would you be able to wrap the testing up on this one today with the checklist?

**Review by pecanoro — DISMISSED** [review]

**Review by mountiny — DISMISSED** [review]
Changes look good to me, @hoangzinh is working on testing this PR and adding checklist now

**Comment by mountiny** [comment]
I have notced two bugs testing on ios native reported [here](https://expensify.slack.com/archives/C036QM0SLJK/p1713196821689339?thread_ts=1713194544.839969&cid=C036QM0SLJK)

https://github.com/Expensify/App/assets/36083550/a4626c20-ad2e-4bfa-bf4d-9ae3093fdcf6


1. The form validation should start after user clicks next so the name inputs should show as wrong only after that
2. I was able to close the modal with gesture (starting to drag down from the Welcome header) and then the app did not work so well, I was not able to open the Concierge chat, the onboarding modal did not start when I signed in with the same account again

File: src/pages/OnboardingPersonalDetails/BaseOnboardingPersonalDetails.tsx
Comment (filip-solecki) [reply src/pages/OnboardingPersonalDetails/BaseOnboardingPersonalDetails.tsx:40]: Basically it focuses `First Name` input here just after transition is ended and everything is set up, doing it by ref instead of `autoFocus={true}` as it was before avoid focus to go to `Composer` in Concierge chat

**Review by hoangzinh — APPROVED** [review]
Overall, LGTM excepts 2 bugs that I reported. All yours @pecanoro @mountiny

**Review by mountiny — DISMISSED** [review]
changes look good to me, thanks @hoangzinh for a thorough and speedy testing ❤️ 

Lets address those comments and we can try to merge this one so its in staging assoon as possible

**Comment by mountiny** [comment]
Making a new build

**Comment by dubielzyk-expensify** [comment]
I don't think we can use elipsis here: 
CleanShot 2024-04-17 at 07 47 03@2x

We'll need to account for translations and further tweak of copy in the future.

So I'd prefer to have the checkmark space built in so that it doesn't jump when it's selected. And for the selection box to be a fixed height but the text to go on two lines if it has to.

**Comment by dubielzyk-expensify** [comment]
Can I also just confirm that the padding to the edge of the modal to the text and the button is `20px`?

CleanShot 2024-04-17 at 07 48 32@2x

**Comment by pecanoro** [comment]
@filip-solecki Is it ready for a final review?

**Comment by cdOut** [comment]
> @filip-solecki Is it ready for a final review?

@pecanoro I'm fairly sure we should still hold off on the final review, as yesterday we've discussed some changes that need to be implemented on the backend before we push this one onto main.

The thread about those backend changes [starts here](https://swmansion.slack.com/archives/C036QM0SLJK/p1713363008121029?thread_ts=1713194544.839969&cid=C036QM0SLJK).

cc: @mountiny, please correct me if I'm wrong.

**Comment by pecanoro** [comment]
Ahh sounds good, I wasn't sure since there were some commits but no message!!

File: src/ONYXKEYS.ts
Comment (mountiny) [inline src/ONYXKEYS.ts:118]:
```suggestion
    /** This NVP contains information about whether the onboarding flow was completed or not */
```

File: src/types/onyx/Onboarding.ts
Comment (mountiny) [inline src/types/onyx/Onboarding.ts:2]:
```suggestion
    /** A Boolean that informs whether the user has completed the guided setup onboarding flow */
```

**Review by mountiny — DISMISSED** [review]
Perfect, in general this is looking good, but we got one problem stemming from the fact the onboarding flow NVP gets updated only in the CompledGuidedSetup API command coming in here https://github.com/Expensify/App/pull/39951

So I think we have to merge that one first and then clean this up and merge the enablement

**Comment by mountiny** [comment]
Noting, I have merged the Stage 2 PR https://github.com/Expensify/App/pull/39951 so we are going to enable it here. 

I have included the test steps from that PR here as well, we will need to merge main, resolve merge conflicts and retest the flow works as expected, especially given the NVP and when the flow is completed the NVP should update cc @filip-solecki @cdOut @rezkiy37 

Also adding @rayane-djouah as C+ from the Stage 2 PR so their experience will be helpful for us to get this over the finish line on Monday

**Comment by mountiny** [comment]
@Expensify/design for last feedback on this before before we get going

**Comment by mountiny** [comment]
Discussing in slack in this [thread](https://expensify.slack.com/archives/C036QM0SLJK/p1713644084643399)

File: src/libs/Navigation/AppNavigator/createCustomBottomTabNavigator/BottomTabBar.tsx
Comment (mountiny) [inline src/libs/Navigation/AppNavigator/createCustomBottomTabNavigator/BottomTabBar.tsx:52]: Now with Stage 2 we have to go to purpose page first
```suggestion
        Welcome.isOnboardingFlowCompleted({onNotCompleted: () => Navigation.navigate(ROUTES.ONBOARDING_PURPOSE)});
```

Comment (filip-solecki) [reply src/libs/Navigation/AppNavigator/createCustomBottomTabNavigator/BottomTabBar.tsx:52]: Yes, and Purpose is set as ROOT so it is all good in this way too

**Comment by filip-solecki** [comment]
I think it is all good here and whoever wants may test 🚀

**Comment by dubielzyk-expensify** [comment]
I get this when I try to tap the checkmark on the task: 
CleanShot 2024-04-22 at 20 05 30@2x

Is there a way to not allow this action at all? 

Also, I tried submitting it in dev but it didn't check as done. Is this because it's not actually working on the back end?

**Comment by dubielzyk-expensify** [comment]
> Also, I tried submitting it in dev but it didn't check as done. Is this because it's not actually working on the back end?

Oh I realized that might be part of a different, PR? If so, apologies!

**Comment by rezkiy37** [comment]
> The backdrop of the modal is not transparent on Web 🤔 Screenshot 2024-04-22 at 17 00 01

Yeah, I will take this problem to fix.

**Comment by dubielzyk-expensify** [comment]
For `Manage my team's expenses`:

**Create a workspace**
CleanShot 2024-04-22 at 20 13 25@2x

**Set up categories:**
CleanShot 2024-04-22 at 20 20 37@2x

**Add expense approvals**
CleanShot 2024-04-22 at 20 24 22@2x

**Invite your team**
CleanShot 2024-04-22 at 20 25 47@2x

**Comment by dubielzyk-expensify** [comment]
I see this in a few places, but basically most of the task comments start with a repeat of the title of the task. If they do, they should be bold

**Comment by dubielzyk-expensify** [comment]
For `Track business spend for taxes`

**Create a workspace**
The first "Create a workspace" string should be bolded

**Track an expense**
CleanShot 2024-04-22 at 20 29 01@2x

**Comment by dubielzyk-expensify** [comment]
For `Get paid back by my employer`:

**Submit an expense**
CleanShot 2024-04-22 at 20 31 33@2x

**Enable your wallet**
CleanShot 2024-04-22 at 20 32 59@2x

**Comment by dubielzyk-expensify** [comment]
For `Track and budget personal spend`

**Track an expense**
CleanShot 2024-04-22 at 20 34 56@2x

File: src/libs/actions/Welcome.ts
Comment (filip-solecki) [reply src/libs/actions/Welcome.ts:62]: Hmm, it seems to be working without them, just wondering if we can completely remove both as we probably does not read them anywhere else

**Comment by dubielzyk-expensify** [comment]
For `Track and budget personal spend`:

(forgot to add it to the pictures, but underlined should be bolded 😅 )

**Start a chat**
CleanShot 2024-04-22 at 20 36 45@2x


**Split an expense**
CleanShot 2024-04-22 at 20 37 52@2x

**Enable your wallet**
CleanShot 2024-04-22 at 20 38 45@2x

**Comment by dubielzyk-expensify** [comment]
Also, the intro video has a piece of text underneath it that says:
> Getting paid is as easy as sending a message.

We'll need to replace that, but I've reached out internally. I'll return with the right copy!

cc @Expensify/design for thoughts too.

File: src/libs/actions/Welcome.ts
Comment (filip-solecki) [reply src/libs/actions/Welcome.ts:62]: @hoangzinh

**Comment by filip-solecki** [comment]
@rezkiy37 do you plan to work on @dubielzyk-expensify's texts related comments as well?

**Comment by rezkiy37** [comment]
> @rezkiy37 do you plan to work on @dubielzyk-expensify's texts related comments as well?

Yeah, it was introduced in the scope of my PR (https://github.com/Expensify/App/pull/39951). So I think it is fair that I should fix them. Also, feel free to ping me if any other bugs come from my changes. Thank you!

**Review by rayane-d — COMMENTED** [review]
LGTM! only issue I see is that the backdrop of the modal is not transparent as reported [here](https://github.com/Expensify/App/pull/39687#issuecomment-2068998775). also, The concierge chat is flickering, I needed to turn on "use staging server" in order to see the messages in concierge chat online.


https://github.com/Expensify/App/assets/77965000/5141917c-836b-4566-a099-0cbc386d8f60

**Comment by trjExpensify** [comment]
@rayane-djouah on the transparent background (or lack of!), were you testing using an adhoc build? @danielrvidal ran into similar a few days ago and @mountiny gave him this possible explanation:

>are you previewing this on an ad hoc build? I _think_ it's a known thing that ad hoc builds usually dump you on a full screen "not found" page, so the modal might just be popping up over that.

**Comment by dubielzyk-expensify** [comment]
Also for this video:

CleanShot 2024-04-23 at 08 19 26@2x

Underneath we want to change the text to be 
> **Welcome to Expensify**
One app to handle all your business and personal spend in a chat. Built for your business, your team, and your friends.
[Get started]

**Comment by danielrvidal** [comment]
@rezkiy37 can you make sure to make the updates asap when you come online tomorrow. This is a high priority one, thank you!

**Review by hoangzinh — COMMENTED** [review]
NAB:

**Comment by filip-solecki** [comment]
It is not reproducible while using staging server, @mountiny is it expected?

**Comment by filip-solecki** [comment]
Hmm, we should probably fix validation to not accept empty strings, right?

**Review by hoangzinh — APPROVED** [review]
Tested again and all goods except this bug https://github.com/Expensify/App/pull/39687#issuecomment-2072846849. I think it's not a blocker if we would like to merge this PR asap. Moreover, it should be fixed from BE, therefore I'm good with this PR. All your @mountiny @pecanoro

**Comment by mountiny** [comment]
> Sorry I mean we can fill in a " " (space) as described in test step 10 in this PR description

I think that instruction does not mean you can fill in name as whitespace, but that we should accept names like `Edgar Alan Poe`

File: src/libs/actions/Welcome.ts
Comment (mountiny) [inline src/libs/actions/Welcome.ts:34]: NAB: Can you please add a comment to some follow up PR to explain that the `onboarding` is an empty array when the user hasnt completed the flow yet

**Review by mountiny — APPROVED** [review]
One NAB, but changes are looking great, lets kick this off

File: src/libs/actions/Welcome.ts
Comment (filip-solecki) [reply src/libs/actions/Welcome.ts:34]: Will add in some follow up

**Comment by Julesssss** [comment]
Hey @pecanoro  @mountiny. FYI deploying this before the Auth PR caused [this issue](https://expensify.slack.com/archives/C049HHMV9SM/p1714470238207479) on prod and staging. It looked pretty bad for new users.

Screenshot 2024-04-30 at 10 40 41

**Comment by pecanoro** [comment]
> FYI deploying this before the Auth PR

What do you mean by this? The PR was tested so I am wondering how these were triggered. Or maybe something that was modified after new small changes were committed?

**Comment by Julesssss** [comment]
I'm not certain, but it looks like [this Auth PR](https://github.com/Expensify/Auth/pull/10696) should have been deployed before the full onboarding changes, as the above bug was deployed to staging/prod. 

Perhaps this PR just modified the onboarding flow and the issue lies with [this issue](https://github.com/Expensify/App/issues/41024), which had the deploy blocker label removed. I'm not sure and don't really have time to investigate further.

**Pull request merged**
